src/zx/FunctionalityConstruction.cpp

```diff
@@ -0,0 +1,425 @@
+
+                case qc::OpType::Z: {
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+                    addCcx(diag, ctrl_0, ctrl_1, target, qubit_vertices);
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+                    break;
+                }
+                default: {
+                    throw ZXException("Unsupported Multi-control operation: " +
+                                      qc::toString(op->getType()));
+                    break;
+                }
+            }
+        } else {
+            throw ZXException("Unsupported Multi-control operation (" + std::to_string(op->getNcontrols()) + " ctrls)" + qc::toString(op->getType()));
+        }
+        return it + 1;
+    }
+
+    ZXDiagram FunctionalityConstruction::buildFunctionality(const qc::QuantumComputation* qc) {
+        ZXDiagram           diag(qc->getNqubits());
+        std::vector<Vertex> qubit_vertices(qc->getNqubits());
+        for (size_t i = 0; i < qc->getNqubits(); ++i) {
+            diag.removeEdge(i, i + qc->getNqubits());
+            qubit_vertices[i] = i;
+        }
+
+        auto initial_layout = qc->initialLayout;
+
+        // if (!initial_layout.empty()) {
+        //     std::vector<Vertex> new_qubit_vertices(qc->getNqubits());
+        //     for (auto i = 0; i < qc->getNqubits(); i++) {
+        //         new_qubit_vertices[i] = qubit_vertices[i];
+        //     }
+        //     for (auto& [tar, src]:
+        //          qc->initialLayout) { // reverse initial permutation
+        //         if (tar == src)
+        //             continue;
+
+        //         auto v_tar = diag.addVertex(tar, 1);
+        //         diag.addEdge(qubit_vertices[src], v_tar);
+        //         new_qubit_vertices[tar] = v_tar;
+        //     }
+        //     qubit_vertices = new_qubit_vertices;
+        // }
+
+        for (auto it = qc->cbegin(); it != qc->cend();) {
+            auto& op = *it;
+
+            if (op->getType() == qc::OpType::Compound) {
+                auto* compOp = dynamic_cast<qc::CompoundOperation*>(op.get());
+                for (auto subIt = compOp->cbegin(); subIt != compOp->cend();)
+                    subIt = parse_op(diag, subIt, compOp->end(), qubit_vertices, qc->initialLayout);
+                ++it;
+            } else {
+                it = parse_op(diag, it, qc->end(), qubit_vertices, qc->initialLayout);
+            }
+        }
+
+        for (size_t i = 0; i < qubit_vertices.size(); ++i) {
+            diag.addEdge(qubit_vertices[i], diag.getOutputs()[i]);
+        }
+        return diag;
+    }
+    bool FunctionalityConstruction::transformableToZX(const qc::QuantumComputation* qc) {
+        for (const auto& it: *qc) {
+            if (!transformableToZX(it.get()))
+                return false;
+        }
+        return true;
```

Comment:
can be replaced by the idiomatic
```suggestion
return std::all_of(qc->cbegin(), qc->cend(), [](const auto& op) { return transformableToZX(op.get()); });
```